projects/components/src/splitter/splitter.service.ts

```diff
@@ -0,0 +1,104 @@
+    direction: SplitterDirection
+  ): Observable<boolean> {
+    const moveOnParent$ = fromEvent<MouseEvent>(parentElement, 'mousemove');
+    const downOnHost$ = fromEvent<MouseEvent>(splitterElement, 'mousedown');
+    const upOnWindow$ = fromEvent<MouseEvent>(window, 'mouseup');
+
+    return downOnHost$.pipe(
+      mergeMap(() => moveOnParent$.pipe(takeUntil(upOnWindow$))),
+      map(event => {
+        const parentBounds = parentElement.getBoundingClientRect();
+        const prevSibling = this.getHTMLElement(splitterElement.previousElementSibling);
+        const nextSibling = this.getHTMLElement(splitterElement.nextElementSibling);
+
+        const previousElementUpdatedDimension = prevSibling
+          ? this.calculateUpdatedDimensionForPreviousElement(event, parentBounds, direction)
+          : 0;
+        const nextElementUpdatedDimension = nextSibling
+          ? this.calculateUpdatedDimensionForNextElement(previousElementUpdatedDimension, parentBounds, direction)
+          : 0;
+        const updatedDimensionData: UpdatedDimensionData[] = [];
+
+        if (prevSibling) {
+          updatedDimensionData.push({
+            element: prevSibling,
+            dimension: previousElementUpdatedDimension
+          });
+        }
+
+        if (nextSibling) {
+          updatedDimensionData.push({
+            element: nextSibling,
+            dimension: nextElementUpdatedDimension
+          });
+        }
+
+        return updatedDimensionData;
+      }),
+      map(data => this.maybeUpdateLayoutOfSiblings(data, direction))
+    );
+  }
+
+  private maybeUpdateLayoutOfSiblings(data: UpdatedDimensionData[], direction: SplitterDirection): boolean {
+    let layoutChanged = false;
+    data.forEach(datum => {
+      if (direction === SplitterDirection.Horizontal) {
+        this.updateWidth(datum.element, datum.dimension);
+      } else {
+        this.updateHeight(datum.element, datum.dimension);
+      }
+      layoutChanged = true;
+    });
+
+    return layoutChanged;
+  }
+
+  private getHTMLElement(element: Element | null): HTMLElement | undefined {
+    return element instanceof HTMLElement ? element : undefined;
+  }
+
+  private calculateUpdatedDimensionForPreviousElement(
+    event: MouseEvent,
+    parentBounds: DOMRect,
+    direction: SplitterDirection
+  ): number {
+    return direction === SplitterDirection.Horizontal
+      ? Math.min(event.clientX - parentBounds.x, parentBounds.width)
+      : Math.min(event.clientY - parentBounds.y, parentBounds.height);
+  }
+
+  private calculateUpdatedDimensionForNextElement(
+    previousElementUpdatedDimension: number,
+    parentBounds: DOMRect,
+    direction: SplitterDirection
+  ): number {
+    return (
+      (direction === SplitterDirection.Horizontal ? parentBounds.width : parentBounds.height) -
+      previousElementUpdatedDimension -
+      4
+    );
+  }
+
+  private updateWidth(element: HTMLElement, width: number): void {
+    element.style.width = `${width}px`;
```

Comment:
in general (for all the dom manipulation/reading) can we use angular's renderer class?

Reply:
I can check that. would it bring this change inside the scope of Change detection?

Reply:
I don't think it would affect change detection (since it's not modifying inputs), and it may not give us any direct benefits right here, but it's a good practice for some reasons that are relevant and some that aren't. In descending order of importance:
1. Angular will do any sanitization (e.g. xss) that we'd normally get in a template but not when we directly modify the DOM.
2. With all changes going through a common renderer, we could in the future choose to link that up to a service like layout change (because we can insert custom renderers).
3. If we ever choose to use something like web workers or server side rendering, direct DOM access would break (while this abstraction would work).

Reply:
okay, let me check

Reply:
I don't see any method to access previous sibling using angular renderer2. I am not sure this would work with the current approach.

Reply:
Yeah, perhaps not. For what it's worth - the approach seems very un-angulary and has some other pitfalls due to that too. It's reaching into the dom outside it and manipulating it, which is broken encapsulation and also makes assumptions on how those things are laid out (for example, if one sibling was using flex, this would break, right?)

The API I'd expect for something like this would be:
```
<ht-splitter [direction]="direction">
   <div *htSplitterContent>Left</div> 
   <div *htSplitterContent>Right</div>
</ht-splitter
```
This would keep everything internal to the splitter.

Reply:
Nope. it woudn't break if it is inside flex container. We could always nest it in various components and directives, but that just add additional structures to DOM. It doesn't do anything else apart from that. Take a deeply nested splitter as an example.

Reply:
It would nest just fine in this example. It does add a dom layer to ensure we're modifying something we own rather than breaking encapsulation, but that seems like a tradeoff I'd take any day.

Reply:
Done

Reply:
Directive and Angular Renderer